packages/loader/container-loader/src/containerStorageAdapter.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    IDocumentStorageService,
+    IDocumentStorageServicePolicies,
+    ISummaryContext,
+} from "@fluidframework/driver-definitions";
+import {
+    ICreateBlobResponse,
+    ISnapshotTree,
+    ISummaryHandle,
+    ISummaryTree,
+    ITree,
+    IVersion,
+} from "@fluidframework/protocol-definitions";
+import { IDisposable } from "@fluidframework/common-definitions";
+import { AttachState } from "@fluidframework/container-definitions";
+
+/**
+ * This class wraps the actual storage and make sure no wrong apis are called according to
+ * container attach state.
+ */
+export class ContainerStorageAdapter implements IDocumentStorageService, IDisposable {
+    private _disposed = false;
+    constructor(
+        private readonly storageGetter: () => IDocumentStorageService,
+        private readonly attachState: () => AttachState,
+        private readonly blobs: Map<string, ArrayBufferLike>,
+    ) {
+    }
+
+    private throwOnNotAttached(name: string) {
+        const attachState = this.attachState();
+        if (attachState !== AttachState.Attached) {
+            throw new Error(`${name} not allowed in Unattached container`);
```

Comment:
Any chance we can also log error to telemetry before we throw? I just want to make sure we catch all such cases.

That said, can we have default behavior in next iteration?
I.e. I assume next move is to remove reading blob content from ISnaphostTree and actually use storage.
And once that happens, this object can not throw on all those calls, but actually can perform as real storage (read blobs, return snapshot, return only single version, etc.).
I can fail writing blobs initial in detached state, but even that we will likely make work (in unrelated work item to support images in drafts)

Reply:
Added telemetry error event. Yes, in next iteration I will put data here in storage.